### PR TITLE
Add tests to validate `react-native-windows-init` options

### LINKED ISSUE #9163: React-native-windows-init should have option validation tests
### Summary

With our other CLI commands, we have both options interfaces and tests which make sure the arguments/options are "valid" - they have descriptions, the proper type, etc.

We should add this to react-native-windows-init

### Motivation

Part of the telemetry refactor (#8253) includes making sure that we make explicit choices about how CLI args are recorded in telemetry. This ensures we proactively strip PII.

While react-native-init is currently fine, adding these tests will make sure the new arguments are implemented properly.

### Basic Example

_No response_

### Open Questions

_No response_

### PR DESCRIPTION
**Pull request by jonthysell**
## Description

This PR makes several improvements to `react-native-windows-init`:

* Created a new WindowsInitOptions interface for better typing
* The telemetry option sanitizer now uses `keyof WindowsInitOptions` to ensure explicit choices around sanitizing PII for new arguments
* The module now exports a `reactNativeWindowsInit` function (instead of just executing) so we can write proper tests
* Added a test to validate the args/options (make sure they have defaults, descriptions and are keys of WindowsInitOptions)
* Fixes a bug where telemetry wasn't picking string args that were undefined

Closes #9163

### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
We have no unit tests for the `react-native-windows-init` package. So developers can introduce bugs when making changes.

Resolves #9163

### What
Added new unit tests. Slightly changed the structure of the Cli module so that it _can_ be unit testable, and added a test to verify the command's args are valid.

## Screenshots
N/A

## Testing
```cmd
cd packages\react-native-windows-init
yarn test
```

I also verified locally that we can still run the command on RN projects.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9174)

### CODE REVIEW

File: packages/react-native-windows-init/src/Cli.ts [hunk 1]
```diff
@@ -42,90 +44,94 @@ const NPM_REGISTRY_URL = validUrl.isUri(npmConfReg)
   ? npmConfReg
   : 'http://registry.npmjs.org';
 
+export const windowsInitOptions: {[key: string]: yargs.Options} = {
```
Comment (NickGerleman) [inline packages/react-native-windows-init/src/Cli.ts:47]: The type given here means that other code can assume that any string is present or not present as a key. So a misspelling of a field like `versionn` would still be accepted in code dealing with it.

```suggestion
export const windowsInitOptions = {
```

If the value of the object literal is stuck in time, you can add [const assertions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) (`as const`) lower down to make the type of `windowsInitOptions` conform to its exact structure.

That doesn't check that each type is a valid yarn option at the declaration-site though (like we had before with this as a param). Only usage-site.

File: packages/react-native-windows-init/src/Cli.ts [hunk 2]
```diff
@@ -335,34 +342,56 @@ function installReactNativeWindows(
...
 
-  const sanitizedOptions = yargsOptionsToOptions(argv, optionSanitizer);
+  const sanitizedOptions = yargsOptionsToOptions(
+    options,
+    optionSanitizer as OptionSanitizer,
+  );
   const sanitizedDefaultOptions = yargsOptionsToOptions(
     yargsParser.parse(''),
-    optionSanitizer,
+    optionSanitizer as OptionSanitizer,
```
Comment (NickGerleman) [inline packages/react-native-windows-init/src/Cli.ts:392]: Why the needed cast here and above?

File: packages/react-native-windows-init/src/Cli.ts [hunk 3]
```diff
@@ -607,5 +639,21 @@ function isProjectUsingYarn(cwd: string): boolean {
     console.error(initWindowsError);
   }
   endTelemetrySession(initWindowsError);
-  setExitProcessWithError(argv.verbose, initWindowsError);
-})();
+  setExitProcessWithError(!!options.verbose, initWindowsError);
+}
+
+export interface WindowsInitOptions {
```
Comment (NickGerleman) [inline packages/react-native-windows-init/src/Cli.ts:645]: Why is this separate from `windowsInitOptions`?

File: packages/react-native-windows-init/src/Cli.ts [hunk 4]
```diff
@@ -575,25 +607,25 @@ function isProjectUsingYarn(cwd: string): boolean {
...
     await generateWindows(process.cwd(), name, ns, {
-      language: argv.language as 'cs' | 'cpp',
-      overwrite: argv.overwrite,
-      verbose: argv.verbose,
-      projectType: argv.projectType as 'lib' | 'app',
-      experimentalNuGetDependency: argv.experimentalNuGetDependency,
-      useWinUI3: argv.useWinUI3,
-      useHermes: argv.useHermes,
+      language: options.language as 'cs' | 'cpp',
+      overwrite: !!options.overwrite,
```
Comment (NickGerleman) [inline packages/react-native-windows-init/src/Cli.ts:619]: The type of `options` looks to be less precise than that of `argv`, which knew this was a never-nullish boolean. `yargs` will infer the type of options based on the values passed into its options params.

File: packages/react-native-windows-init/src/Cli.ts [hunk 5]
```diff
@@ -607,5 +639,21 @@ function isProjectUsingYarn(cwd: string): boolean {
...
+  setExitProcessWithError(!!options.verbose, initWindowsError);
+}
+
+export interface WindowsInitOptions {
+  version?: string;
+  namespace?: string;
+  verbose?: boolean;
+  telemetry?: boolean;
+  language?: 'cpp' | 'cs';
+  overwrite?: boolean;
```
Comment (NickGerleman) [inline packages/react-native-windows-init/src/Cli.ts:651]: `overwrite`, and several other options are never undefined. Being either required, or having a default (which yargs will populate).

**Comment by jonthysell** [comment]
@NickGerleman Rather than spreading this all out across your comments:

For our other CLI commands, we have two things - an object defining our options for Commander, and an interface of what we _expect_ to be extracted that we use within our command. The process looks like:

1. We give the options object to the RN CLI
2. The core RN CLI code passes that to Commander, which parses the args into a resolved object
3. The CLI then calls our command function with that object (which we arbitrarily declare in our method signature is of the interface type, but technically that isn't enforced by anything, which is why all of our interfaces have to use optional fields for everything)

For validation we have tests for each CLI command which verify that the options object is valid (name, description, etc) and that our interface contains fields for each option. With this setup, we can also do the `keyof` screening in the options sanitizer function to make sure that we _explicitly_ define sanitization behavior to strip PII.

All together, if people try to add new options to a command, they have to add it correctly everywhere otherwise they'll get build or test failures.

Now I'm trying to do the same for `react-native-windows-init`. Instead of calling `yargs.options` with the object inline, I created a const of that object (`windowsInitOptions`). I also created the interface `WindowsInitOptions`, switched to `keyof WindowsInitOptions` in the sanitizer, and created the validation test. Since it's not part of the RN CLI, we do each part ourselves:

1. We give the options object to yargs
2. We call yargs to parse our args into a resolved object
3. We cast that resolved object `as WindowsInitOptions`

Unfortunately, yargs has some quirks. It can only infer option types when passed an options-object inline (i.e. `yargs.options({ all the keys and values inline}`) not when passed the `windowsInitOptions` object. That's why `argv` had all the right types that made TS happy. If we return to passing in the object inline to yargs, then I don't know how to extract that metadata back out of yargs in the validation tests.

**Comment by NickGerleman** [comment]
@jonthysell I took a look at the `yargs` type definitions. I think the following should work to maintain correct type information:

```ts
const windowsInitOptions = <const>{
  version: {
    type: 'string',
    describe: 'The version of react-native-windows to use.',
  },
  namespace: {
    type: 'string',
    describe:
      "The native project namespace. This should be expressed using dots as separators. i.e. 'Level1.Level2.Level3'. The generator will apply the correct syntax for the target language",
  },
  //...
};

type InitOptions = yargs.InferredOptionTypes<typeof windowsInitOptions>;
```

**Comment by NickGerleman** [comment]
We can also do this, to ensure anything wrong in the options structure is caught early:

```ts
const windowsInitOptions = initOptions(<const>{
  version: {
    type: 'string',
    describe: 'The version of react-native-windows to use.',
    required: true,
  },
  namespace: {
    type: 'string',
    describe:
      "The native project namespace. This should be expressed using dots as separators. i.e. 'Level1.Level2.Level3'. The generator will apply the correct syntax for the target language",
  },
  //...
});

// Causes the type-checker to ensure the options object is a valid yargs options object
function initOptions<T extends Record<string, yargs.Options>>(options: T): T {
  return options;
}

type WindowsInitOptions = yargs.InferredOptionTypes<typeof windowsInitOptions>;
```

File: packages/react-native-windows-init/src/Cli.ts (hunk 2 above)
Comment (jonthysell) [reply packages/react-native-windows-init/src/Cli.ts:392]: Changing to the keyof for the key param breaks this for some reason.

**Comment by jonthysell** [comment]
> We can also do this, to ensure anything wrong in the options structure is caught early:
> 
> ```ts
> const windowsInitOptions = initOptions({
>   version: {
>     type: 'string',
>     describe: 'The version of react-native-windows to use.',
>     required: true,
>   },
>   namespace: {
>     type: 'string',
>     describe:
>       "The native project namespace. This should be expressed using dots as separators. i.e. 'Level1.Level2.Level3'. The generator will apply the correct syntax for the target language",
>   },
>   //...
> });
> 
> // Causes the type-checker to ensure the options object is a valid yargs options object
> function initOptions<T extends Record<string, yargs.Options>>(options: T): T {
>   return options;
> }
> 
> type WindowsInitOptions = yargs.InferredOptionTypes<typeof windowsInitOptions>;
> ```

I have implemented this. It has the benefit of cleaning up the nullable booleans, but now changed the WindowsInitOptions type to allow any strings for enum values. I've updated the tests to compensate slightly by verifying when a default isn't in the list of choices.

File: packages/react-native-windows-init/src/Cli.ts [hunk 6]
```diff
@@ -460,21 +490,24 @@ function isProjectUsingYarn(cwd: string): boolean {
   return !!findUp.sync('yarn.lock', {cwd});
 }
 
-(async () => {
-  if (argv.verbose) {
-    console.log(argv);
+export async function reactNativeWindowsInit(args?: string[]) {
+  args = args ?? process.argv;
```
Comment (asklar) [inline packages/react-native-windows-init/src/Cli.ts:494]: if using process.argv, don't we need to shift 1 or 2 slots to get rid of the `node.exe` and the `scriptname.js` params?

File: packages/react-native-windows-init/src/e2etest/cli.test.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    // Validate description
+    expect(option.describe).toBeDefined();
+    expect(option.describe).not.toBeNull();
+    expect(option.describe!).toBe(option.describe!.trim());
+
+    // Validate all command options are present in WindowsInitOptions
+    expect(
+      validateOptionName(optionName, optionName as keyof WindowsInitOptions),
+    ).toBe(true);
+  }
```
Comment (asklar) [inline packages/react-native-windows-init/src/e2etest/cli.test.ts:71]: should we be testing that passing an option that isn't defined results in the value not being present in the sanitized version?

**Review by asklar — APPROVED** [review]

File: packages/react-native-windows-init/src/Cli.ts (hunk 6 above)
Comment (jonthysell) [reply packages/react-native-windows-init/src/Cli.ts:494]: No, yargsParser.parse cleans it up.

File: packages/react-native-windows-init/src/e2etest/cli.test.ts (hunk 7 above)
Comment (jonthysell) [reply packages/react-native-windows-init/src/e2etest/cli.test.ts:71]: yargsParser.parse will fail the command if an unsupported arg is passed in, and display the help, before any of our (telemetry) code is run.

**Pull request merged**